### PR TITLE
add very generic concat algorithm

### PR DESCRIPTION
**Pull request by arnej27959**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

@lesters please review
@havardpe please review

### CODE REVIEW

**Review by lesters — APPROVED** [review]
LGTM!

File: vespajlib/src/main/java/com/yahoo/tensor/functions/Concat.java [hunk 1]
```diff
@@ -27,6 +31,233 @@
...
+                return 1;
+            }
+            if (sizes.size() == 1) {
+                return sizes.iterator().next();
+            }
+            throw new IllegalArgumentException("inconsistent size of concat dimension, had "+sizes.size()+" different values");
+        }
+    }
+
+    static class Helper {
```
Comment (havardpe) [inline vespajlib/src/main/java/com/yahoo/tensor/functions/Concat.java:86]: consider `ConcatPlan`

File: vespajlib/src/main/java/com/yahoo/tensor/functions/Concat.java [hunk 2]
```diff
@@ -27,6 +31,233 @@
...
+        final String concatDimension;
+
+        SideHow aHow = new SideHow();
+        SideHow bHow = new SideHow();
+
+        enum CombineHow { left, right, both, concat }
+
+        List<CombineHow> combineHow = new ArrayList<>();
+        
+        TensorAddress combine(TensorAddress match, TensorAddress leftOnly, TensorAddress rightOnly, int concatDimIdx) {
```
Comment (havardpe) [inline vespajlib/src/main/java/com/yahoo/tensor/functions/Concat.java:97]: consider moving 'out'

File: vespajlib/src/main/java/com/yahoo/tensor/functions/Concat.java [hunk 3]
```diff
@@ -27,6 +31,233 @@
...
+            while (j < bDims.size()) {
+                bOnly(bDims.get(j++).name());
+            }
+            if (combineHow.size() < resultType.rank()) {
+                var idx = resultType.indexOfDimension(concatDimension);
+                combineHow.add(idx.get(), CombineHow.concat);
+            }
+        }
+
+        Tensor mergeSides(SideInfo a, SideInfo b) {
```
Comment (havardpe) [inline vespajlib/src/main/java/com/yahoo/tensor/functions/Concat.java:195]: consider moving 'out'

File: vespajlib/src/main/java/com/yahoo/tensor/functions/Concat.java [hunk 4]
```diff
@@ -27,6 +31,233 @@
...
+
+    enum DimHow { match, expand, concat }
+
+    static class SideHow {
+        List<DimHow> handleDims = new ArrayList<>();
+        int numMatch = 0;
+        int numExpand = 0;
+    }
+
+    SideInfo analyse(Tensor side, SideHow how) {
```
Comment (havardpe) [inline vespajlib/src/main/java/com/yahoo/tensor/functions/Concat.java:229]: consider `decompose`

**Review by havardpe — APPROVED** [review]
the code looks good
consider some minor name/structure tweaks

**Comment by arnej27959** [comment]
@havardpe please take another look

**Review by havardpe — APPROVED** [review]

**Pull request merged**